docs/proposals/policy.md

```diff
@@ -0,0 +1,438 @@
+  1.  proxy
+  1.  permission-request
+  1.  * - represents all verbs
+
+### Verb Specific Attributes
+Different verbs will require different attributes.  For instance, a list would have a kind, a get would have an kind and an id, and an update may have a kind, an id, and a list of the fields being modified.  Although exec isn't well defined, it could very well have an entirely distinct set of attributes.  The exact contract rules will be chosen by the Authorizer/AuthorizationAttributeBuilder pair, but basic resource access for openshift should include kind in order to be compatible with existing kubernetes authorization.
+
+#### Why allow GroupNames instead of simply saying that RoleBindings only have a list of UserNames?  
+Ease of maintenance and separation of powers.  Groups span namespaces, but RoleBindings do not.  If you specify the full set of users on every RoleBinding, then you have to repeat it for every namespace.  If that membership changes, you now have to update it in multiple locations.  As for the separation of powers, if you have rights to modify a RoleBinding, you have the power to change its permissions (RoleTemplateName) instead of only having the power to change its membership.  It seems reasonable to allow someone to modify group membership ("chubba is mechanic") without also having the ability to modify permissions associated with the group itself ("mechanics are super-admins").  Having Groups distinct from Roles and RoleBindings makes it easy to express this in policy.
+
+## API
+### Basic CRUD
+The basic CRUD for Roles and RoleBindings, and Policy is not as straightforward as usual.  
+
+The REST API will respond to gets, lists, creates, updates, and deletes (not watches) at the expected locations for Roles and RoleBindings.  The REST API will respond to gets, lists, and watches (not creates, updates, and deletes) at the expected location for Policy.  
+
+Only one Policy document is allowed per namespace.  Roles and RoleBindings are not stored in etcd.  Instead, requests against their REST endpoints will result in inspection and modification of the Policy document.  This makes it easier to express restrictions on how policy can be manipulated, while making it possible to guarantee of referential integrity for references inside the local namespace.
+
+#### allowedentities?namespace=<namespace>
+**Please suggest a better name and a better location to put this** Given a namespace which users and groups can see or edit some resource contained in that namespace.  This API is not guaranteed to return accurate results.  It is possible that a user returned in the list, could be denied access to any resource in the namespace.  This can happen when a user is granted allow rights for something the namespace, but a deny rule later stomps on it.  Barring weird cases like that, it will work.  A user must have get permission on requested namespace's policy object in order to make this call.  The response will look like this:
```

Comment:
You can still cache these decisions if necessary, but I agree POST is more appropriate than GET if only because we can version it effectively.


> On Jan 17, 2015, at 1:25 AM, Derek Carr <notifications@github.com> wrote:
> 
> In docs/proposals/policy.md:
> 
> > +### Verb Specific Attributes
> > +Different verbs will require different attributes.  For instance, a list would have a kind, a get would have an kind and an id, and an update may have a kind, an id, and a list of the fields being modified.  Although exec isn't well defined, it could very well have an entirely distinct set of attributes.  The exact contract rules will be chosen by the Authorizer/AuthorizationAttributeBuilder pair, but basic resource access for openshift should include kind in order to be compatible with existing kubernetes authorization.
> > +
> > +#### Why allow GroupNames instead of simply saying that RoleBindings only have a list of UserNames?  
> > +Ease of maintenance and separation of powers.  Groups span namespaces, but RoleBindings do not.  If you specify the full set of users on every RoleBinding, then you have to repeat it for every namespace.  If that membership changes, you now have to update it in multiple locations.  As for the separation of powers, if you have rights to modify a RoleBinding, you have the power to change its permissions (RoleTemplateName) instead of only having the power to change its membership.  It seems reasonable to allow someone to modify group membership ("chubba is mechanic") without also having the ability to modify permissions associated with the group itself ("mechanics are super-admins").  Having Groups distinct from Roles and RoleBindings makes it easy to express this in policy.
> > +
> > +## API
> > +### Basic CRUD
> > +The basic CRUD for Roles and RoleBindings, and Policy is not as straightforward as usual.  
> > +
> > +The REST API will respond to gets, lists, creates, updates, and deletes (not watches) at the expected locations for Roles and RoleBindings.  The REST API will respond to gets, lists, and watches (not creates, updates, and deletes) at the expected location for Policy.  
> > +
> > +Only one Policy document is allowed per namespace.  Roles and RoleBindings are not stored in etcd.  Instead, requests against their REST endpoints will result in inspection and modification of the Policy document.  This makes it easier to express restrictions on how policy can be manipulated, while making it possible to guarantee of referential integrity for references inside the local namespace.
> > +
> > +#### allowedentities?namespace=<namespace>
> > +**Please suggest a better name and a better location to put this** Given a namespace which users and groups can see or edit some resource contained in that namespace.  This API is not guaranteed to return accurate results.  It is possible that a user returned in the list, could be denied access to any resource in the namespace.  This can happen when a user is granted allow rights for something the namespace, but a deny rule later stomps on it.  Barring weird cases like that, it will work.  A user must have get permission on requested namespace's policy object in order to make this call.  The response will look like this:
> Please suggest a better name and a better location to put this?
> 
> I think a more standard term for this type of operation is a resource-centric access review.
> 
> From the resource, find all permissions over it; then, for each permission, determine all the roles that have those permissions; and, finally, find all subjects that have (or can activate) any of those roles.
> 
> I actually think I would prefer this API to be a POST that takes as input a ResourceAccessReview which says for a given resource reference (kind / name / uid) in a namespace (foo), return me a ResourceAccessReviewResponse (i.e. the list of of subjects - i.e. users or groups) that can access it.
> 
> This means I could make ResourceAccessReview requests for any resource type in a namespace - not just the namespace. This is more powerful because I can say tell me who can see this Namespace/Project, or I can say, tell me who can see this particular pod, service, etc.
> 
> I think in this manner, the API has more broad applicable use, and does not really impact implementation cost.
> 
> curl -X POST /api/{version}/ns/{namespace}/resourceaccessreviews -d @resource-access-review.json
> Request / Response objects
> 
> type ResourceAccessReview struct {
>   TypeMeta
>   Resource ObjectReference
> }
> 
> type ResourceAccessReviewResponse struct {
>   TypeMeta
>   Subjects []Subject
> }
> 
> type Subject struct {
>   Name string
>   Type string // i.e. is this a user or group
> }
> This pattern keeps the API resource-centric, and avoids a desire to feel like you can cache the response of the GET style request pattern you presented.
> 
> —
> Reply to this email directly or view it on GitHub.
>